### PR TITLE
Refresh expired Capture access_tokens

### LINKED ISSUE #35: Capture access tokens expire and then the widget gets messed up
to fix:
- switch to `response_type` of `code`.
- exchange that code for a token on the server side
- store the refresh token
- use `janrain.capture.ui.createCaptureSession` to feed the token to the widget

### PR DESCRIPTION
**Pull request by ErinCall**
Fixes #35

### CODE REVIEW

File: lib/user.rb [hunk 1]
```diff
@@ -48,6 +57,26 @@ def self.from_hash(hash)
...
+    get_token(grant_type: 'code', code: code)
+  end
+
+  def self.get_token(post_args)
+    body = {
+      redirect_uri: CAPTURE_URL,
+      client_id: CAPTURE_OWNER_CLIENT_ID,
+      client_secret: CAPTURE_OWNER_CLIENT_SECRET,
+    }.merge(post_args)
+    response = HTTParty.post(CAPTURE_URL+'/oauth/token', {body: body})
```
Comment (ErinCall) [inline lib/user.rb:95]: Let's put some memcaching on here so we're not _utterly_ reliant on Capture being up...

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -7,12 +7,23 @@ class ApplicationController < ActionController::Base
 
   helper_method :user_signed_in?
   before_filter :load_janrain_facts
+  before_filter :current_user
```
Comment (ErinCall) [inline app/controllers/application_controller.rb:16]: this is kind've a funky hack, and while I think it needs to be done, it probably merits a comment saying "here's what's going on with this funky hack:"

File: lib/user.rb [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
+  CACHE_TTL = 5 * 60 #seconds
   attr_accessor :meetup_token, :email, :confirmed_at, :uuid, :name, :profile_photo_url
 
   def self.fetch_from_token(token)
-    response = HTTParty.post(CAPTURE_URL + '/entity', {body:{
-      access_token: token,
-      type_name: 'user',
-    }})
+    Rails.cache.fetch("user_token:#{token}", expires_in: CACHE_TTL) do
+      response = HTTParty.post(CAPTURE_URL + '/entity', {body:{
```
Comment (peterkeen) [inline lib/user.rb:9]: Ruby scoping is mystifying. I would have expected `response` to be in a local scope inside the block. Guess not!

Comment (peterkeen) [inline lib/user.rb:9]: Oh wait I read the rest of the code lol never mind lol

File: lib/user.rb (hunk 1 above)
Comment (peterkeen) [inline lib/user.rb:95]: Are you going to do this thing?

Comment (ErinCall) [inline lib/user.rb:95]: ...I put this comment in the wrong place. This isn't a call that a cache can be correct about. The thing I cached was the correct thing to cache.

Comment (peterkeen) [inline lib/user.rb:95]: Oh ok.

**Comment by peterkeen** [comment]
:thumbsup: :ship: :it:

**Pull request merged**